src/Datadog.Trace/DiagnosticListeners/DiagnosticObserver.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.DiagnosticListeners
+{
+    internal abstract class DiagnosticObserver : IObserver<KeyValuePair<string, object>>
+    {
+        private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.For<DiagnosticObserver>();
+
+        protected DiagnosticObserver(IDatadogTracer tracer)
+        {
+            Tracer = tracer ?? throw new ArgumentNullException(nameof(tracer));
+        }
+
+        protected IDatadogTracer Tracer { get; }
+
+        public virtual bool IsSubscriberEnabled()
+        {
+            return true;
+        }
+
+        public virtual IDisposable SubscribeIfMatch(DiagnosticListener diagnosticListener)
+        {
+            if (diagnosticListener.Name == GetListenerName())
+            {
+                return diagnosticListener.Subscribe(this, IsEventEnabled);
+            }
+
+            return null;
+        }
+
+        void IObserver<KeyValuePair<string, object>>.OnCompleted()
+        {
+        }
+
+        void IObserver<KeyValuePair<string, object>>.OnError(Exception error)
+        {
+        }
+
+        void IObserver<KeyValuePair<string, object>>.OnNext(KeyValuePair<string, object> value)
+        {
+            try
+            {
+                OnNext(value.Key, value.Value);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Event Exception: {0}", value.Key);
+            }
+        }
+
+        /// <summary>
+        /// Gets the name of the <see cref="DiagnosticListener"/> that should be instrumented.
+        /// </summary>
+        /// <returns>The name of the <see cref="DiagnosticListener"/> that should be instrumented.</returns>
+        protected abstract string GetListenerName();
```

Comment:
nit: Not sure what coding convention we all prefer, but this could be a good use of an abstract property instead of a traditional Getter

Reply:
This line came directly from [`opentracing-contrib/csharp-netcore`](https://github.com/opentracing-contrib/csharp-netcore/blob/ee64533c0257a88c75f2d2c9759bfabc37243d56/src/OpenTracing.Contrib.NetCore/Internal/DiagnosticListenerObserver.cs#L17), but I agree with you.

Changed in afd24c95e6ffbf68bbb37ace2849ed48bfd0adfe.